osu.Game/Tests/Visual/TestPlayer.cs

```diff
@@ -48,6 +49,20 @@ public TestPlayer(bool allowPause = true, bool showResults = true, bool pauseOnF
             PauseOnFocusLost = pauseOnFocusLost;
         }
 
+        protected override void PrepareReplay()
+        {
+            var autoplayMod = Mods.Value.OfType<ModAutoplay>().FirstOrDefault();
+
+            // This logic should really not exist (and tests should be instantiating a ReplayPlayer), but a lot of base work is required to make that happen.
+            if (autoplayMod != null)
+            {
+                DrawableRuleset?.SetReplayScore(autoplayMod.CreateReplayScore(GameplayBeatmap.PlayableBeatmap, Mods.Value));
+                return;
+            }
+
+            base.PrepareReplay();
```

Comment:
What is all of this? Why? Isn't this handled by `DrawableRuleset` as well (see my comment above)?

Reply:
Is the entire point of all of this actually just the `return` statement, in order to block replay recording/scoreprocessor binding?

I still don't understand what the rest of this is doing - it's setting the replay score which is also done by `DrawableRuleset` too so there's code duplication that I can't understand. It can't exist in two places like this.

Reply:
It's intentionally using `ModAutoplay`. I explain this here: https://github.com/ppy/osu/pull/12952#issuecomment-851858594

Reply:
That comment doesn't explain why this can't use the `ICreateReplay` interface?

Reply:
Well I see two solutions:

- Remove the interface and use `ModAutoplay` everywhere
- Add... an inline comment mentioning we explicitly want autoplay here? Because that's what we want right? Not some other mod which creates a replay in the future. The logic is only here as a stand-in until tests are (maybe?) migrated to using different base test scenes (ie `ReplayPlayerTestScene` and `PlayerTestScenes. I hope I explained well enough why this is actually a lot of work. I have a branch showing how large the change becomes if you'd like me to push it.

Reply:
Here's the changes, which mostly work but the case I mentioned above is where I stopped going down this path: https://github.com/ppy/osu/commit/a6d590b8145cebc883d0aac2dd185a1ec830d396

Reply:
Fine, keep this as it is. I don't really agree that it should be tied to modautoplay when it's a hack for the fact that all testscenes create this class, but I don't care that much to continue this discussion.